gluonnlp/data/stream.py

```diff
@@ -453,3 +388,117 @@ def _write(data, target, buffers, seq_len, i, length):
         return
+
+
+class _Prefetcher(object):
+    """Interal shared prefetcher logic."""
+    data_queue = None
+    control_queue = None
+
+    def __init__(self, stream, num_prefetch, seed, np_seed, mx_seed):
+        super(_Prefetcher, self).__init__()
+        self.stream = stream
+        assert num_prefetch > 0, 'Unbounded Prefetcher is unsupported.'
+        self.num_prefetch = num_prefetch
+        self.seed = seed
+        self.np_seed = np_seed
+        self.mx_seed = mx_seed
+
+    def run(self):
+        """Method representing the process’s activity."""
+        random.seed(self.seed)
+        np.random.seed(self.np_seed)
+        mx.random.seed(self.mx_seed)
+
+        stream_iter = iter(self.stream)
+        while True:
+            try:  # Check control queue
+                c = self.control_queue.get(False)
+                if c is None:
+                    break
+            except queue.Empty:
+                pass
+
+            try:
+                data = next(stream_iter)
+                self.data_queue.put(data)
+            except StopIteration:
+                self.data_queue.put(None)
+
+    def __next__(self):
+        next_item = self.data_queue.get()
+        if next_item is None:
+            self.control_queue.put(None)
+            raise StopIteration
+        return next_item
+
+    def next(self):
+        return self.__next__()
+
+    def __iter__(self):
+        return self
+
+
+class _ProcessPrefetcher(_Prefetcher, multiprocessing.Process):
+    """Interal multi-processing prefetcher."""
+
+    def __init__(self, *args, **kwargs):
+        super(_ProcessPrefetcher, self).__init__(*args, **kwargs)
+        self.data_queue = multiprocessing.Queue(self.num_prefetch)
+        self.control_queue = multiprocessing.Queue()
+        self.daemon = True
+        self.start()
+
+
+class _ThreadPrefetcher(_Prefetcher, threading.Thread):
+    """Interal threaded prefetcher."""
+
+    def __init__(self, *args, **kwargs):
+        super(_ThreadPrefetcher, self).__init__(*args, **kwargs)
+        self.data_queue = queue.Queue(self.num_prefetch)
+        self.control_queue = queue.Queue()
+        self.daemon = True
+        self.start()
+
+
+class PrefetchingStream(object):
+    """Prefetch a DataStream in a separate Thread or Process.
+
+    This iterator will create another thread or process to perform
+    ``iter_next`` and then store the data in memory. It potentially accelerates
+    the data read, at the cost of more memory usage.
+
+    The python, numpy and mxnet random states in the launched Thread or Process
+    will be initialized randomly based on the next 32 bit integer in the
+    python, numpy and mxnet random generator of the caller respectively
+    (random.getrandbits(32), numpy.random.randint(0, 2**32),
+    int(mx.nd.random.uniform(0, 2**32).asscalar())).
+
+    Parameters
+    ----------
+    stream : DataStream
+        Source stream.
+    num_prefetch : int, default 1
```

Comment:
is `num_prefetch ` > 1 also supported?

Reply:
It should be. num_prefetch sets the size of the queue. The Thread / Process tries to push as many elementsas possible to the queue.

Reply:
I updated the tests to run with different values of num_prefetch as well as thread / process.

Reply:
Oh okay I was under the wrong impression that this is the number of prefetching threads. nvm